README.rst
iDoneThis CLI Utility
=====================

|Version| |Documentation| |Status| |Coverage| |License| |GPA|

A simple, small, and opinionated Python CLI for interacting with the awesome
iDoneThis service in a way that suits for my workflow.

This CLI allows you to record your grandiose accomplishments of the day did
today as well as see what others on your team did.

You can find more docs on `ReadTheDocs <http://idonethis.readthedocs.org/en/latest/>`_.


Record what you've done
-----------------------

::

  $ idonethis -m 'Holy smoke I did it!' --token 'my-auth-token' --team 'backend'
  Recorded what you've done, keep up the good work!

  $ echo 'Holy smoke I did it!' | idonethis --token 'my-auth-token' --team 'backend'
  Recorded what you've done, keep up the good work!


Or if you're feeling fancy and want to use your ``$EDITOR``

::

  $ idonethis --token 'my-auth-token' --team 'backend'
  Recorded what you've done, keep up the good work!


See what others have done
-------------------------

::

  $ date
  Thu Jun  4 19:10:11 EDT 2015

  $ idonethis summary --token 'my-auth-token' --team 'backend'
  The "backend" team did this on 2015-06-04

  DanT
  ----

  * Made a PR on something
  * Created a little CLI for stuff

  OtherGuy
  --------

  * Fixed all the things


Or if you want to see who was slacking yesterday

::

  $ idonethis summary --date yesterday --token 'my-auth-token' --team 'backend'
  The "backend" team did this on 2015-06-04

  DanT
  ----

  * Thought about doing work, then didn't

  OtherGuy
  --------

  * Planning to fix all the things


Installation
------------

You can install this off of PyPI using PIP.

::

    $ pip install idonethis


.. include:: contributing.rst


Development
-----------

Python 2.7, 3.2, 3.3, 3.4, and Pypy 2.1 are all supported and integrated
against.  To run `detox`_ locally you'll need all the interpreters... or you
can do what I do and throw it over the fence to TravisCI and hope my config
file hasn't broken again.

1. Create a new virtual environment
2. Install development requirements from *requirements.txt*
3. Run tests  ``nosetests``
4. `detox`_ is installed and will run the test suite across all supported python platforms
5. `python setup.py build_sphinx` will generate documentation into *build/sphinx/html*

TL;DR
+++++

::

    $ virtualenv env
    $ ./env/bin/pip install -qr requirements.txt
    $ source env/bin/activate
    (env) $ nosetests
    (env) $ python setup.py build_sphinx
    (env) $ detox


.. include:: ../HISTORY.rst


License
-------

MIT

.. include:: ../LICENSE


.. _detox: https://testrun.org/tox/

.. |Version| image:: https://badge.fury.io/py/idonethis.svg?
   :target: http://badge.fury.io/py/idonethis

.. |Status| image:: https://travis-ci.org/djt5019/idonethis.svg?branch=master
   :target: https://travis-ci.org/djt5019/idonethis

.. |Coverage| image:: https://coveralls.io/repos/djt5019/idonethis/badge.svg
  :target: https://coveralls.io/r/djt5019/idonethis

.. |Downloads| image:: https://pypip.in/d/idonethis/badge.svg?
   :target: https://pypi.python.org/pypi/idonethis

.. |License| image:: http://img.shields.io/:license-mit-blue.svg
   :target: http://doge.mit-license.org

.. |Documentation| image:: https://readthedocs.org/projects/idonethis/badge/?version=latest
   :target: http://idonethis.readthedocs.org/en/latest
   :alt: Documentation Status

.. |GPA| image:: https://codeclimate.com/github/djt5019/idonethis/badges/gpa.svg
   :target: https://codeclimate.com/github/djt5019/idonethis
   :alt: Code Climate


HISTORY.rst
.. :changelog:

Release History
===============

`1.0.0`_ (2015-06-04)
---------------------

* Initial release with ability to add a new done, no summary functionality yet.
* Add documentation and coverage/testing/ci support.
* Add console script so the CLI is uable after being PIP installed.
* Rename the repo from iDoneIt to match the iDoneThis API name.



.. _`1.0.0`: https://github.com/djt5019/idonethis/compare/0.0.0...1.0.0


tests.py
import os
import unittest

import betamax
import idonethis
import requests

try:
    from unittest import mock
except ImportError:
    import mock


# No requests should come from Travis
RECORD_MODE = 'never' if bool(os.environ.get('TRAVIS')) else 'once'
AUTH_TOKEN = 'Token ' + os.environ.get('IDONETHIS_TOKEN', 'x' * 20)


class BetamaxMixn(object):

    @classmethod
    def setUpClass(cls):
        super(BetamaxMixn, cls).setUpClass()
        cls.session = requests.Session()
        cls.session.headers = {'Authorization': AUTH_TOKEN}
        cls.vcr = betamax.Betamax(cls.session)

        with betamax.Betamax.configure() as config:
            config.cassette_library_dir = 'fixtures'
            config.default_cassette_options['record_mode'] = RECORD_MODE
            config.define_cassette_placeholder('Token <AUTH_TOKEN>',
                                               AUTH_TOKEN)


class TestSubmitDone(BetamaxMixn, unittest.TestCase):

    def test_invalid_auth_token(self):
        self.session.headers['Authorization'] = 'Token hunter2'

        with self.assertRaises(ValueError):
            with self.vcr.use_cassette('bad_auth_token'):
                idonethis.submit_done(self.session, 'team', 'done')

        self.session.headers['Authorization'] = AUTH_TOKEN

    def test_using_the_wrong_team(self):
        with self.assertRaises(ValueError):
            with self.vcr.use_cassette('bad_team'):
                idonethis.submit_done(self.session, 'team', 'done')

    def test_a_good_request(self):
        with self.vcr.use_cassette('good_post_request'):
            idonethis.submit_done(self.session, 'aweber-be-bof', 'I did it!')


class TestGrabText(unittest.TestCase):

    @mock.patch('sys.stdin')
    def test_stdin_available(self, stdin):
        stdin.isatty.return_value = False
        stdin.read.return_value = 'testing'
        self.assertEqual('testing', idonethis.get_done_text())

    @mock.patch('idonethis.subprocess.call')
    def test_bringing_up_an_editor(self, subprocess_call):
        self.assertEqual('Today I did... ', idonethis.get_done_text())


class TestMainFunction(BetamaxMixn, unittest.TestCase):

    @mock.patch('sys.exit')
    @mock.patch('sys.stdin')
    @mock.patch('idonethis.parser')
    @mock.patch('idonethis.requests')
    def test_no_done_text_provided_in_cli(self, requests, parser, stdin, exit_):
        requests.Session.return_value = self.session
        parser.parse_args.return_value = mock.Mock(
            message=None,
            team='aweber-be-bof',
            token=os.environ.get('IDONETHIS_TOKEN', 'x' * 20))

        stdin.isatty.return_value = False
        stdin.read.return_value = 'testing'

        with self.vcr.use_cassette('good_post_request'):
            idonethis.main()

        self.assertFalse(exit_.called)

    @mock.patch('idonethis.requests')
    @mock.patch('idonethis.parser')
    @mock.patch('idonethis.get_done_text')
    def test_done_text_provided_in_cli(self, get_done_text, parser, requests):
        requests.Session.return_value = self.session
        parser.parse_args.return_value = mock.Mock(
            message='I did it!',
            team='aweber-be-bof',
            token=os.environ.get('IDONETHIS_TOKEN', 'x' * 20))

        with self.vcr.use_cassette('good_post_request'):
            idonethis.main()

        self.assertFalse(get_done_text.called)

    @mock.patch('sys.exit')
    @mock.patch('idonethis.parser')
    @mock.patch('idonethis.requests')
    def test_unhandled_exception_occurs(self, requests, parser, exit_):
        requests.Session.return_value = self.session
        parser.parse_args.return_value = mock.Mock()

        with self.vcr.use_cassette('good_post_request'):
            idonethis.main()

        self.assertTrue(exit_.called)


idoneit.py
# -*- coding: utf-8 -*-
version_info = (1, 0, 0)
__version__ = '.'.join(str(v) for v in version_info[:3])

import os
import subprocess
import sys
import tempfile

import argparse
import requests


ROOT_URI = 'https://idonethis.com/api/v0.1/dones/'

parser = argparse.ArgumentParser()
parser.add_argument('-m', '--message', help='Content of your done')
parser.add_argument('--team', required=True, help='Your team')
parser.add_argument('--token', required=True,
                    help='Your authorization token: see https://idonethis.com/api/token/')


def get_done_text():
    """Grab input from either the users editor of choice or stdin."""
    if not sys.stdin.isatty():
        return sys.stdin.read()

    EDITOR = os.environ.get('EDITOR', 'vim')

    with tempfile.NamedTemporaryFile(suffix='.tmp', delete=False) as temp:
        filename = temp.name
        temp.write(b'Today I did... ')
        temp.flush()
        subprocess.call([EDITOR, temp.name])

    with open(filename) as updated_temp:
        data = updated_temp.read()

    os.unlink(filename)
    return data


def submit_done(session, team, done):
    """Submit the users done and bask in the glory of productivity.

    :param Session session: A `requests_` session object.
    :param str team: Name of the team to the user did something on.
    :param str done: Description of what was accomplished.

    """
    post_data = {'raw_text': done, 'team': team}
    response = session.post(ROOT_URI, json=post_data)

    if response.ok:
        return

    if response.status_code not in (400, 401):
        response.raise_for_status()

    payload = response.json()

    if 'errors' in payload and 'team' in payload['errors']:
        raise ValueError("You aren't a part of the '{}' team".format(team))
    else:
        raise ValueError(payload['detail'])


def main(args):
    """The main entry point for the application."""
    token = args.token
    team = args.team
    done_text = args.message

    # Seed the Authorization headers for all subsequent requests
    session = requests.Session()
    session.headers['Authorization'] = 'Token {}'.format(args.token)

    if not done_text:
        done_text = get_done_text()

    submit_done(session, team, done_text)


if __name__ == '__main__':  # pragma: nocover
    try:
        main(parser.parse_args())
        print("Recorded what you've done, keep up the good work!")
    except Exception as exception:
        exit("Failed to record what you've done: {}".format(exception))


idonethis.py
# -*- coding: utf-8 -*-
version_info = (1, 0, 0)
__version__ = '.'.join(str(v) for v in version_info[:3])

import os
import subprocess
import sys
import tempfile

import argparse
import requests


ROOT_URI = 'https://idonethis.com/api/v0.1/dones/'

parser = argparse.ArgumentParser()
parser.add_argument('-m', '--message', help='Content of your done')
parser.add_argument('--team', required=True, help='Your team')
parser.add_argument('--token', required=True,
                    help=('Your authorization token: '
                          'see https://idonethis.com/api/token/'))


def get_done_text():
    """Grab input from either the users editor of choice or stdin."""
    if not sys.stdin.isatty():
        return sys.stdin.read()

    EDITOR = os.environ.get('EDITOR', 'vim')

    with tempfile.NamedTemporaryFile(suffix='.tmp', delete=False) as temp:
        filename = temp.name
        temp.write(b'Today I did... ')
        temp.flush()
        subprocess.call([EDITOR, temp.name])

    with open(filename) as updated_temp:
        data = updated_temp.read()

    os.unlink(filename)
    return data


def submit_done(session, team, done):
    """Submit the users done and bask in the glory of productivity.

    :param Session session: A `requests_` session object.
    :param str team: Name of the team to the user did something on.
    :param str done: Description of what was accomplished.

    """
    post_data = {'raw_text': done, 'team': team}
    response = session.post(ROOT_URI, json=post_data)

    if response.ok:
        return

    if response.status_code not in (400, 401):
        response.raise_for_status()

    payload = response.json()

    if 'errors' in payload and 'team' in payload['errors']:
        raise ValueError("You aren't a part of the '{}' team".format(team))
    else:
        raise ValueError(payload['detail'])


def main():
    """The main entry point for the application."""
    args = parser.parse_args()
    token = args.token
    team = args.team
    done_text = args.message

    # Seed the Authorization headers for all subsequent requests
    session = requests.Session()
    session.headers['Authorization'] = 'Token {}'.format(token)

    if not done_text:
        done_text = get_done_text()

    try:
        submit_done(session, team, done_text)
    except Exception as exception:
        sys.exit("Failed to record what you've done: {}".format(exception))

    print("Recorded what you've done, keep up the good work!")


if __name__ == '__main__':  # pragma: nocover
    main()
